### PR TITLE
[PD-2215] Upgrade nodejs

### PR DESCRIPTION
**Pull request by darrint**
Once this is merged we will require npm version 3 and likely nodejs 5.7.x.

Nodejs 5.7.x is already installed and working on the build machine.

Notably npm-shrinkwrap is a smaller file.

### CODE REVIEW

File: gulp/gulpfile.js [hunk 1]
```diff
@@ -54,16 +41,11 @@ function webpackConfig() {
       reporting: './src/reporting/main',
       manageusers: './src/manageusers/main',
       nonuseroutreach: './src/nonuseroutreach/main',
-      vendor: vendorLibs,
```
Comment (darrint) [inline gulp/gulpfile.js]: Originally this was supposed to speed up the development build. It no longer makes a difference and the build is simpler without it.

File: gulp/gulpfile.js [hunk 2]
```diff
@@ -54,16 +41,11 @@ function webpackConfig() {
...
     },
     resolve: {
       root: path.resolve('src'),
       // you can now require('file') instead of require('file.coffee')
       extensions: ['', '.js', '.jsx'],
-      // Thank you internet man! http://stackoverflow.com/a/32444088
-      alias: {
-        'react': path.join(__dirname, 'node_modules', 'react'),
```
Comment (darrint) [inline gulp/gulpfile.js]: This is no longer needed with npm 3. Originally react-bootstrap bundled (??!) react, resulting in parallel versions.

File: dk2/dev/Dockerfile [hunk 3]
```diff
@@ -1,17 +1,21 @@
...
+run curl -s https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add -
+run apt-get update
+
 run apt-get --yes build-dep python-mysqldb libgeos-dev && apt-get clean
 run apt-get --yes install python-setuptools python-dev apache2 python-pip \
-    mysql-client libxml2 libxml2-dev libxslt-dev python-mysqldb npm \
-    phantomjs python-ipdb strace && apt-get clean
-
-# Node.js
+    mysql-client libxml2 libxml2-dev libxslt-dev python-mysqldb \
```
Comment (aspidites) [inline dk2/dev/Dockerfile]: I thought there was a recent conversation saying that python-mysqldb was already in requirements.txt, and thus didn't need/shouldn't be installed explicitly by apt?

File: doc/source/developers/styleguides/javascript.rst [hunk 4]
```diff
@@ -125,6 +125,15 @@ Library Policy
   * Are we bringing in a large library only to use a very small part of it?
     (using a tank where a stick would do)
 
+* Avoid editing ``package.json`` by hand. Use ``npm install --save`` and
+  other npm commands as much as possible. These automatically keep the
+  shrinkwrap file up to date.
+
+* NPM likes to put a lot of unstable information in the shrinkwrap file.
+  If you find a lot extraneous lines in the diff of `npm-shrinkwrap.json`,
```
Comment (aspidites) [inline doc/source/developers/styleguides/javascript.rst]: Missing a preposition here?

File: gulp/gulpfile.js (hunk 1 above)
Comment (aspidites) [inline gulp/gulpfile.js]: Why does this no longer make a difference? What's the consequence of this no longer making a difference? Are produced bundles now larger than they were before? Smaller?

File: gulp/gulpfile.js [hunk 5]
```diff
@@ -129,7 +112,9 @@ gulp.task('bundle', function(callback) {
       callback('webpack error');
       return;
     }
-    fs.writeFile('profile.json', JSON.stringify(stats.toJson(), null, 4));
+    if (produceWebpackProfile) {
```
Comment (aspidites) [inline gulp/gulpfile.js]: THANK YOU!!!

File: gulp/clean-shrinkwrap.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Run this script to clean the npm-shrinkwrap.json file if it fills with
+ * stray "resolved" and "from" items.
+ *
+ * $ npm run clean-shrinkwrap
+ *
+ * based on "shonkwrap"
+ * https://github.com/skybet/shonkwrap/blob/master/shonkwrap
```
Comment (tdphillips) [inline gulp/clean-shrinkwrap.js]: I'm a little concerned by this line. It looks like the relation is a bit closer than "based on".

File: gulp/gulpfile.js (hunk 1 above)
Comment (darrint) [inline gulp/gulpfile.js]: For the production build, commons plugin doesn't need this help.

For development, this was my attempt to get webpack to do less work for incremental builds under gulp watch. Now webpack-dev-server handles incrementally building and does so very fast. (<1s on my machine. This is subject to verification on an OSX machine. I've been surprised before.)

With the addition of the `minChunks: 2` configuration in webpack.config.js we get larger vendor.js and smaller app bundles.

File: dk2/dev/Dockerfile (hunk 3 above)
Comment (darrint) [inline dk2/dev/Dockerfile]: I missed that conversation but it sounds right. I'll try taking it out and make sure things still work in the image. (It appeared twice.)

File: gulp/clean-shrinkwrap.js (hunk 6 above)
Comment (darrint) [inline gulp/clean-shrinkwrap.js]: What should it say? They aren't identical.

Comment (tdphillips) [inline gulp/clean-shrinkwrap.js]: It's been modified to no longer be a bash script, no longer use require.js, and have slightly different indentation. I guess you're right - they're not identical.

Comment (aspidites) [inline gulp/clean-shrinkwrap.js]: "adapted from"?

**Pull request merged**